src/rgw/rgw_auth_keystone.cc

```diff
@@ -179,6 +184,33 @@ TokenEngine::get_acl_strategy(const TokenEngine::token_envelope_t& token) const
       }
     }
 
+    for (const auto& r : token_roles) {
+      if (r.is_reader) {
+        if (r.is_admin) {    /* system scope reader persona */
```

Comment:
What is the difference between an admin and an admin reader? Is the admin reader supposed to only perform read ops? In that case even if you assign RGW_OP_TYPE_READ perm to the admin reader, I think, it will be able to perform write operations also because of https://github.com/ceph/ceph/blob/master/src/rgw/rgw_process.cc#L230.

Reply:
The whole "Safe/Secure RBAC" project came about because RBAC was too flexible and operators often made mistakes assigning roles. So, now so-called "personas" are defined. They are essentially models which come with canned roles. Originally there were 9 of them: admin, auditor (reader), and member types for each of system, domain, and project scopes. Truly an abomination only an architect could create. As we started to implement it, it was pared down, and all that's left are readers, and only system and project ones at that, because domain scope was "postponed". Personally, I think that project reader is completely useless, but the management dug the heels in. That's why Swift has them split into two commits. At first I hoped they'd get get off my back if only system reader were implemented.

Fundamentally readers are worsened, or "safed" admins: they are admins that can only read. As such, they defeat ordinary permissions. And, because system and project admins exist, so do system and project readers. System reader can access any data in the cluster, but project reader can only do it within Keystone project (which used to be called "tenant").